### PR TITLE
Add Dockerfile based on alpine:3.9

### PR DESCRIPTION
**Pull request by mawi12345**
This Dockerfile creates a 762MB sized image based on the alpine python3 image.

### CODE REVIEW

**Comment by mawi12345** [comment]
Please note that the @testing repository of alpine linux is currently used for the py3-pybind11 and jbig2enc packages.
If you are interested, I can extend the pull request and build the package in the image.

File: .docker/alpine.dockerfile [hunk 1]
```diff
@@ -0,0 +1,24 @@
+FROM python:3-alpine
+
+RUN \
+  echo '@testing http://nl.alpinelinux.org/alpine/edge/testing' >> /etc/apk/repositories \
+  # Add runtime dependencies
+  && apk add --update jbig2enc@testing ghostscript qpdf \
+    tesseract-ocr unpaper libxml2-dev libxslt-dev zlib-dev \
+    qpdf-dev py3-pybind11@testing leptonica-dev libffi-dev binutils \
```
Comment (jbarlow83) [inline .docker/alpine.dockerfile]: pngquant should be added

**Comment by jbarlow83** [comment]
Basically my preference would be to replace the existing Dockerfiles with Alpine since it's often a better fit for Docker, or at least it has the reputation. I'm not actually that familiar with it. The fact that it has most things packaged is quite nice - dependency support is better than any other platform. 

As it currently stands I wouldn't accept this because I don't see the point of an alternative image that from the user standpoint should be functionally identical. But I'd like to enlist your help with establishing whether Alpine really is a better fit, and if so, migrating all of the Docker images to Alpine.

Does Alpine repackage system Python packages like Debian? (They have their reasons but it's quite inconvenient for upstream.)

Are you willing to help out with transition all of the Docker files to Alpine?

**Comment by mawi12345** [comment]
Alpine Linux docker images are smaller than Ubuntu / Debian images. In my opinion, this is the only and great advantage. The ocrmypdf-alpine image (762MB) is compared to the the current ocrmypdf-ubuntu image (1.39GB) 45% smaller. 

In the alpine package repositories many [python projects are packaged](https://pkgs.alpinelinux.org/packages?name=py3*&branch=v3.9).

But I would like to base the ocrmypdf-alpine image of [python:3-alpine](https://github.com/docker-library/python/blob/6c42dbe19061cb1abcda541dc0c7a4d0aef5aacf/3.8-rc/alpine3.9/Dockerfile) and do not use python packets from alpine repositories (pybind11 would also have to be built). Then the docker build is independent of python dependencies from the alpine py3-* packages.

If you replace the ocrmypdf-ubuntu image with the ocrmypdf-alpine image that would break images that are based on e.g. jbarlow83/ocrmypdf-webservice.

I can imagine very well that there are users who are using the
jbarlow83/ocrmypdf-webservice image in the following way:

```
FROM jbarlow83/ocrmypdf-webservice

# Do something that will not work on alpine
RUN apt-get update && apt-get install ...

COPY my-webservice.py /application/webservice.py

...
```

That's why I'd at least temporarily maintain both docker images.

I'm not certain how much sense it makes to maintain multiple images with different tesseract languages. 
I would only provide english in the default image.

It is very simple to create a docker image with the desired languages:

```
FROM jbarlow83/ocrmypdf-alpine

# Add german
RUN apk add tesseract-ocr-data-deu
```

[tesseract-ocr-data-* alpine packages](https://pkgs.alpinelinux.org/packages?page=1&branch=v3.9&name=tesseract-ocr-data%2A&arch=x86)

It may be sufficient to have documentation available on how to build an image with your desired language(s)?

And i would like to help to transition / extend the docker files =)

In the meantime  I have found some other improvements to the above docker file that can reduce the resulting image size.

I would suggest that you close this pull request and we move the discussion to an issue?

**Comment by mawi12345** [comment]
FYI: The docker image build from the alpine:3.9 image (commit 447a377) is only 345MB 👍

**Comment by jbarlow83** [comment]
This is looking very good.

What's the reason for excluding .git from the build? We could save the space, and if included it could capture unreleased branches.

>But I would like to base the ocrmypdf-alpine image of [python:3-alpine](https://github.com/docker-library/python/blob/6c42dbe19061cb1abcda541dc0c7a4d0aef5aacf/3.8-rc/alpine3.9/Dockerfile) and do not use python packets from alpine repositories (pybind11 would also have to be built). Then the docker build is independent of python dependencies from the alpine py3-* packages.

Sounds reasonable.

**Comment by mawi12345** [comment]
> What's the reason for excluding .git from the build? We could save the space, and if included it could capture unreleased branches.

The pip installation does not work without the .git directory. I did not look into the details but i am pretty sure the version of the package is extracted by something like `git describe`.

The app directory (including .git) is not captured in the second build stage (only the final wheel). So it does not enlarge the final image.

I moved away from using python:3-alpine because the image is too large. The current version is based on the bare alpine:3.9 image and also does not use python packets from alpine repositories. All python dependencies are installed via pip.

If it's okay with you i would:

- [ ] Add multi lang image
- [ ] Add webservice image
- [ ] Squash commits

**Comment by mawi12345** [comment]
Since 85fbf77 the source code and test suite is included in the image (+8MB)

Build: `docker build -f .docker/alpine.dockerfile . -t mawi12345/ocrmypdf`
Run: `ocker run mawi12345/ocrmypdf`
Test: `docker run -it --workdir /app --entrypoint python3 mawi12345/ocrmypdf setup.py test`

Test results: 2 failed, 170 passed, 26 skipped, 1 xfailed, 2 warnings
failed: 

- test_bad_locale: `No such file or directory: \'locale\': \'locale\'\n'`? 
- test_french (OK french is not installed)

**Comment by jbarlow83** [comment]
This is shaping up nicely, thanks for working on it.

It occurs to me we could copy webservice.py script into the Docker image as well. Webservice users would just specify an alternate `--entrypoint`.

And we could get rid of the polyglot image too. The default languages can be English, German, and Chinese Simplified - my users seem to be one of those. Then use your instructions for adding languages to the others.

I will keep the old Docker images around but change the documentation to point people to the new ones.

Regarding `test_bad_locale` - this is my workaround for https://bugs.python.org/issue28180, which is fixed in Python 3.7 but present in 3.6. It's probably a good idea to set the locale to `LANG=C.UTF-8` anyway, even in 3.7. This issue describes seems to have some information about setting up locale on Alpine:
https://github.com/gliderlabs/docker-alpine/issues/144#issuecomment-436455850

You're right the .git failure would be tied to `setuptools_scm`. I think that the change you made to install the .egg-info/PKGINFO bit would fix this as setuptools_scm will check other sources of information if .git is not available. 

I will also change `test_french` to test for German.

**Comment by mawi12345** [comment]
Pulling `tesseract-ocr-data-deu`, `tesseract-ocr-data-chi_sim` and `flask` in the image increased the size from 353MB to 417MB. In my opinion, the advantage of maintaining only one image is greater than the disadvantage that the image is 64MB larger.

Run: `docker run mawi12345/ocrmypdf`

Webservice: `docker run --entrypoint python3 -p 5000:5000 mawi12345/ocrmypdf webservice.py`

Tests: `docker run --entrypoint python3 mawi12345/ocrmypdf setup.py test`

**Comment by mawi12345** [comment]
The test session just finished an reported the same failures.

The `bad_locale` test still fails but i am not sure if the image is affected by https://bugs.python.org/issue28180.

When execute the following in the container:

```
/app # python3 -c "import sys; print(sys.getfilesystemencoding())"
utf-8
```

The encoding is set to `utf-8`?!?

**Comment by jbarlow83** [comment]
Issue 28180 applies because it's a Python 3.6 issue.

I think you just need ``ENV LANG=C.UTF-8`` in the Dockerfile. If I'm not mistaken [this](https://bugs.alpinelinux.org/issues/7374) indicates Alpine supports C.UTF-8 out of box.

**Comment by mawi12345** [comment]
I'm not sure how t deal with this Issue28180. I have added `ENV LANG=C.UTF-8` in de08c65. 
Nevertheless the test session reports the failed `test_bad_locale` test. Can we ignore the test in the alpine build? I really don't wont to pull full glibc local support in the image to create the locale file if this issue is a false positive test.

**Comment by jbarlow83** [comment]
I checked it out and I agree `test_bad_locale` is a false positive.

**Pull request merged**